x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/SecondaryAuthenticator.java

```diff
@@ -76,7 +88,10 @@ public void authenticateAndAttachToContext(RestRequest request, ActionListener<S
         // Use cases for secondary authentication are far more likely to want to fall back to the primary authentication if no secondary
         // auth is provided, so in that case we do no want to set anything in the context
         authenticate(
-            authListener -> authenticationService.authenticate(request, false, authListener),
+            authListener -> authenticationService.authenticate(request, false, authListener.delegateFailure((l, authentication) -> {
+                auditTrailService.get().authenticationSuccess(request);
```

Comment:
I agree it's easier to follow, but looking closer I don't think this is going to work correctly.

`auditTrailService.get().authenticationSuccess(request);` looks in the thread context for the authentication, but the `SecondaryAuthenticator` makes it such that the thread context doesn't contain the secondary authN when it invokes the listener of `authenticateAndAttachToContext`.

Reply:
It is possible to have `AuditTrail.authenticationSuccess` taking an extra argument which specifies what key the authentication object is stored against, e.g.
```
void authenticationSuccess(RestRequest request, String authcContextKey);
```

or have it take a Function that knows how to extract authentication, e.g.
```
void authenticationSuccess(RestRequest request, Function<ThreadContext, Authentication> authcReader);
```

This would allow us to always call `authenticationSuccess` in the filter. It is not critical. I'll leave it to you to make the call. Thanks!

Reply:
Hmmm... yeah, I haven't thought of that.
Though if `authenticationSuccess` audits the "current" authentication passing a selector doesn't look elegant to me.
Alternatively, we could change the "authenticationSuccess` method to audit all the authentications, but now it's getting "smarter" and it means we'll to wait for both to complete before we invoke the method. A less smarter option is to pass the authentication or the secondary authentication as arguments, instead of using a no-arg that looks at the "current" authn.

Overall I think there isn't much difference between all of these options. I think I'll leave it to what's there right now: ie. a simple method no-arg method, that audits the "current" authentication, but that needs to be called in the `SecondaryAuthenticator` too.